remote ollama dev environment (#195)
opentelekomcloud/docs-next: docs/blueprints/by-use-case/ai/securely-expose-remote-ollama-endpoints-to-your-developmentd-machine.md

---

> id: securely-expose-remote-ollama-endpoints-to-your-developmentd-machine

Reviewer: Why you called it developmentd and not development ?

title: Securely Expose Remote Ollama Endpoints to your Development Machine

> tags: [ollama, vpn]

Reviewer: maybe add also ai tag here ?

- **Cost Efficiency** : Local testing reduces reliance on cloud resources for every test or development cycle, which can lead to cost savings, especially in scenarios requiring frequent updates and tests.
- **Customized Environment Testing** : Exposing endpoints locally allows you to create a controlled environment that mirrors your production setup. This capability ensures that any AI model behavior is consistent with expectations before deploying broadly.
- **Latency Reduction** : By running services on your local machine, you can significantly reduce latency compared to accessing cloud-hosted endpoints, which enhances the speed and responsiveness of development tools and testing scripts.
- **Compliance and Regulation Adherence** : Certain industries have stringent compliance requirements for data processing and storage. Running AI models locally helps adhere to these regulations by ensuring that data does not leave your secure environment without proper safeguards.
  
By securely exposing Ollama endpoints, you can achieve a balance between operational efficiency, security, and compliance while facilitating a robust development process for AI models on your machine.

:::caution

> Be aware that this Blueprint can incur additional costs, concerning ingress and egress data.

Reviewer: ```suggestion
Be aware that this blueprint may incur additional costs related to ingress and egress data.
```
sounds somehow better

:::

## Prerequisites

For this Blueprint, we are going to need:

> 1. an **ECS Server (Ubuntu 22.04)**: a GPU-accelerated instance (8 vCPUs, 32 GiB) `pi2.2xlarge.4` will suffice.

Reviewer: ```suggestion
1. an **ECS server (Ubuntu 22.04)**: a GPU-accelerated instance (8 vCPUs, 32 GiB) `pi2.2xlarge.4` will suffice.
```

2. a **Point-to-Site VPN connection**: We need to establish a connection between our development machine and the VPC hosting the Ollama VM.
3. an **Ollama** instance: Ollama must be installed on the ECS server above.

:::important
If you don't currently have an Ollama setup, please refer to the official guide for both manual and automated installation options available at this [link](https://github.com/ollama/ollama/blob/main/docs/linux.md).

Make sure you [**add Ollama as a startup service**](https://github.com/ollama/ollama/blob/main/docs/linux.md#adding-ollama-as-a-startup-service-recommended).
:::

## Creating a Point-to-Site VPN

Exposing an Ollama endpoint by assigning an Elastic IP (EIP) directly to a Virtual Machine or using Destination Network Address Translation (DNAT) through a NAT Gateway is considered dangerous due to several key concerns. First, assigning an EIP makes the VM accessible from the internet, significantly increasing its exposure and risk of unauthorized access. This direct accessibility enlarges the attack surface, leaving it vulnerable to potential breaches.

> Additionally, having such an endpoint exposed can make it a target for Distributed Denial of Service (DDoS) attacks aimed at overwhelming your service with excessive traffic, potentially causing downtime or degraded performance. Moreover, using DNAT via a NAT Gateway does not eliminate these risks entirely; it still requires meticulous configuration and management of security groups and firewall rules to ensure only legitimate traffic is allowed. Misconfigurations in this setup can easily lead to unintentional exposure.

Reviewer: ```suggestion
Additionally, having such an endpoint exposed can make it a target for Distributed Denial of Service (DDoS) attacks aimed at overwhelming your service with excessive traffic, potentially causing downtime or degraded performance. Moreover, using DNAT via a NAT Gateway does not eliminate these risks entirely; it still requires meticulous configuration and management of security groups and firewall rules to ensure only legitimate traffic is allowed. Misconfigurations in such setups can easily result in unintentional exposure.
```

sounds smoother

## Changing Ollama's Listening Address

Let's inspect the network sockets associated with Ollama service:

```shell
sudo ss -tupln | grep ollama

tcp   LISTEN   0   4096   127.0.0.1:11434   0.0.0.0:*   users:(("ollama",pid=2728253,fd=3))
```

We instantly notice that although Ollama is bound to all available network interfaces (`0.0.0.0`) and listens on any port (`*`), in practice, it restricts incoming connections to `127.0.0.1:11434`. `127.0.0.1` is the loopback IP address (localhost), meaning **it's accessible only from the same machine**. The port `11434` is where connections should be made to communicate with Ollama. That will naturally prohibit us accessing the Ollama endpoint from our development machine, or any other machine for that matter. What we need to do is to instruct Ollama service to allow incoming connection to any (or even better to a specific interface). For that matter we need to:

1. Stop temporarily the Ollama service:

    ```shell
    sudo systemctl stop ollama
    ```

2. Change the systemd service unit for Ollama to accept incoming connections to all interfaces:

    ```shell
    sudo nano /etc/systemd/system/ollama.service
    ```

>     and add to it an additionall environment variable: `OLLAMA_HOST=0.0.0.0:11434`

Reviewer: ```suggestion
    and add to it an additional environment variable: `OLLAMA_HOST=0.0.0.0:11434`
```

    [Service]
    ExecStart=/usr/local/bin/ollama serve
    User=ollama
    Group=ollama
    Restart=always
    RestartSec=3
    Environment="PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin"
    Environment="OLLAMA_HOST=0.0.0.0:11434"

    [Install]
    WantedBy=default.target
    ```

3. Reload systemd configuration files:

    ```shell
    sudo systemctl daemon-reload
    ```

4. Start the Ollama service on your system

    ```shell
    sudo systemctl start ollama
    ```

## Creating Security Group

1. Go to *Network Console* -> *Security Groups*, and click *Create Security Group*. Add two new **Inbound Rules**:
   - **Protocol**: `TCP`
   - **Port**: `11434`
   - **Source**: `192.168.10.0/24`

:::caution
Although we configured Ollama service to listen to any interface, we want to restrict access to the Ollama endpoint **only** to resources that reside inside the same VPC **and not to anyone that might potentially reach this VM**, `0.0.0.0/0`.

The VPC CIDR for this lab was `192.168.10.0/24`, this might differ in your environment given your individual configuration of the VPC and its Subnets so adjust the **Source** of the Inbound Rule accordingly.

> After establishing the P2S VPN connection, your development machine will be technically a part of this VPC and will be allowed to access the Ollama endpoint.

Reviewer: ```suggestion
Once the P2S VPN connection is established, your development machine will technically be part of this VPC and will be able to access the Ollama endpoint.
```
sounds smoother

:::

2. Add the new Security Group to the Security Groups of the ECS Server.

## Validation

> Check connectivity from your development machine with cURL:

Reviewer: ```suggestion
Verify connectivity from your development machine using cURL:
```

:::important

> - Requires that you have already established the VPN connection.

Reviewer: ```suggestion
- Ensure that the VPN connection is already established.
```

- `192.168.10.183` is the private IP address assigned by the DHCP server of our VPC to the ECS Server that Ollama is installed on. Remember from the step before that the VPC CIDR for this lab was `192.168.10.0/24`, this might differ in your environment given your individual configuration of the VPC and its Subnets.

:::

```shell

> curl http://192.168.10.183:11434/api/tags | jq

Reviewer: ```suggestion
curl http://192.168.10.183:11434/api/tags
```

jq is just pretty print and needs to be installed on most systems as its no standard software
